### ISSUE TITLE
Indicate interpolation type by color

### ISSUE THREAD

**Issue by morevnaproject** [opentoonz/opentoonz#3339]
I am proposing to improve the representation of keyframe-based animation in XSheet/Timeline.

Currently by looking on XSheet you can't say for sure which type of interpolation is used.

![screenshot_006](https://user-images.githubusercontent.com/332868/84243345-44d5e280-ab2c-11ea-9433-d76d135aeaa6.png)

I suggest to change color for line, which connecting keyframes.
The color will indicate type of interpolation.

![screenshot_004](https://user-images.githubusercontent.com/332868/84243415-5c14d000-ab2c-11ea-8784-21352a0956ad.png)

I am suggesting following colors:
- :red_circle: - Constant
- :green_circle: - Speed In/Speed Out
- :yellow_circle: - Linear Interpolation
- :large_blue_circle: - Ease In/Ease Out
- :purple_circle: - Exponential
- :orange_circle: - Expression
- :white_circle: - File

**Comment by DarrenTAnims** [comment]
I don't think I'd remember the colours, to recognise the interpolation type, **BUT** It'd be obvious if you had different colours if something was wrong. Also, it looks like you've done so in your mockup, but if this does get implemented, could you make the line wider, like the coloured one is, here in the mockup? It's really thin and difficult to right click on. And maybe make the click area 1 extra pixel to either side of the line, just for ease of clicking?

**Comment by morevnaproject** [comment]
> could you make the line wider, like the coloured one is, here in the mockup? It's really thin and difficult to right click on. And maybe make the click area 1 extra pixel to either side of the line, just for ease of clicking?

I agree, this would be absolutely great! But I guess it is better to submit as separate issue? ^__^

> I don't think I'd remember the colours

When interpolation is indicated with colors it is so much easier to "read" what movement is going on. Synfig has this for ages and this is a big help:
![screenshot_007](https://user-images.githubusercontent.com/332868/84248546-caa95c00-ab33-11ea-8515-825fc1865da1.png)

(although, the interpolation system in Synfig works a bit differently, this is why it colors key points, not intervals). ^__^

**Comment by RodneyBaker** [comment]
As much as I like the idea of adding color, color for the interpolations will get confused with the (hopefully pending) change to the Curve Editor where the channels are colored.  That coloring is fairly standard in both 2D and 3D programs with x (Red), y (green) and z (blue).  Aside:  white is often used/reserved for the w channel (quaternions, etc.).

The colored channels can be see in Jeremy (turtletooth) Bullock's Tahoma fork.
It looks like this:
![color channels](https://user-images.githubusercontent.com/2576432/84262184-c4d75a80-aae2-11ea-95c9-10c8b7e5bb97.png)


I do like the general idea of finding a way to express the type of interpolation at a quick or immediate glance.

**Comment by morevnaproject** [comment]
> As much as I like the idea of adding color, color for the interpolations will get confused with the (hopefully pending) change to the Curve Editor where the channels are colored

I don't think this can lead to confusion as those are elements from different windows. ^__^

**Comment by RodneyBaker** [comment]
Yes, but that window also has interpolations, keys etc.
So as the UI elements related to interpolations continue to evolve eventually it will result in confusion.

**Comment by morevnaproject** [comment]
Sorry, but I couldn't agree. ^__^ Using color for one element of interface shouldn't forbid to use it in some other element of interface.

In Blender colors used for keyframes and for curves and this not a problem for anyone.
![screenshot_008](https://user-images.githubusercontent.com/332868/84265149-90988400-ab4c-11ea-95a0-14516c3af63f.png)

**Comment by gab3d** [comment]
I think it is a good thing to be able to tell from the Xsheet what is the interpolation type of each segment, and I support the idea and efforts towards that goal.

One thing I'd like to point out though, is that from a UX point of view, the use of colors as the only method to tell between such a variety of interpolation options will be of no use to "color blind" users. And from that standpoint, it'd be good to also think of another visual cues to cover as much users as possible.

I know that OT's system of defining the interpolation for a segment and not keys does away with the idea of using what other known programs have been using for that matter, but anyway I think it'd be good if at least we could brainstorm possible alternate solutions for this.

**Comment by morevnaproject** [comment]
> will be of no use to "color blind" users.

At the moment there are no indication at all, so from this point of view we can say that currently all OT users are "color blind". :rofl: 

My idea is that it is better to use at least something and then (as second step) think how we can improve that for color blind people.

**Comment by gab3d** [comment]
@morevnaproject 
> 
> At the moment there are no indication at all (...)

I know... I know... I'm just saying that it'd be good to take these things into consideration also.

Perhaps we can find a simple way to show that with some kind of small icon in the middle of the colored line? The icon would change to reflect the type of interpolation of that segment.

We could also make it so that left clicking on that little icon it showed a drop down menu to change the interpolation of that segment... it could be more intuitive than what we have now, even. 🤔

**Comment by Lovewyrm** [comment]
red_circle - Constant
    green_circle - Speed In/Speed Out
    yellow_circle - Linear Interpolation
    large_blue_circle - Ease In/Ease Out
    purple_circle - Exponential
    orange_circle - Expression
    white_circle - File

How about the key symbol, and, in a tiny font:
Co
Si/So
Li
Ei/Eo
En
Xp
Fi

Next to it.
Colors are wayyyyyyyy to important to be wasted on this, in my humble opinion. 
Colors should be used to really make data pop. Not to necessarily differentiate a data type.

**Comment by RodneyBaker** [comment]
If color is pursued here's a consideration that comes to mind with regard to what colors should be assigned to which type of interpolation.

While not universal, the color red tends to be representative of an error.
Constant would likely work well as black... which would also make a good default for many cases; a solid foundation, immovable and solid.

Outside of the modern age of color themes, color assignment is largely arbitrary (although cultural as well) but when considering color the question to ask might be:  "in what way does the color red represent a constant value?".  We might suggest that red represents 'stopping', as with a red stop sign but I would argue the color is more representative of the danger of not stopping.  Likewise the color yellow which is used for many/most warning signs.  The content of the signs is one thing but the color standing out is what has led to yellow being used for warning signs.

As I've stated from the outset, I have no real issue with adding color but do have concerns about which colors are assigned.  This is simply because color communicates.

So... money where the mouth is... if were were to assign colors what might I suggest?

Black - Constant
White - Linear
Shades of Green and/or Blue - Spline  (this includes all the various interpolations such as Ease in/out)
Violet/Purple - Exponential  
Light Grey - Expression
Dark Grey - File

This to present basic 'families' of interpolations.
On the range of grayscale from black to white we find:   Constant -- Expression -- File -- Linear
On the range between Blue and Green we find:  Spline interpolations (ease/speed/zero slope, etc.)
At the color Purple we find:  Exponential

This would give users three basic families of color to consider and commit to memory.
Four categories of color if we separate black/white (for constant/linear) from grayscale (expression/file).

Of course these would be the default colors but if allowed via Preferences users could adjust them to any color the desired.

**Comment by RodneyBaker** [comment]
Not quite the same as the request but some progress can be seen in this area in the Ztoryc fork.
Please refer to the release notes a this link:  https://github.com/matitanimata/ztoryc/releases/tag/v0.10.0

UI preferences can often be contentious but discussing such now can get us closer to cocensus as we approach v2.0
